go/vt/vttablet/tabletmanager/vreplication/table_plan_builder.go

```diff
@@ -471,33 +491,80 @@ func (tpb *tablePlanBuilder) generateUpdate(buf *sqlparser.TrackedBuffer, bvf *b
 		separator = ", "
 		switch cexpr.operation {
 		case opExpr:
-			if after {
-				bvf.mode = bvAfter
-				buf.Myprintf("%v", cexpr.expr)
-			} else {
-				buf.WriteString("null")
-			}
+			buf.Myprintf("values(%s)", cexpr.colName.String())
 		case opCount:
-			switch {
-			case before && after:
-				buf.Myprintf("%s", cexpr.colName.String())
-			case before:
-				buf.Myprintf("%s-1", cexpr.colName.String())
-			case after:
-				buf.Myprintf("%s+1", cexpr.colName.String())
-			}
+			buf.Myprintf("%s+1", cexpr.colName.String())
+		case opSum:
+			buf.Myprintf("%s", cexpr.colName.String())
+			buf.Myprintf("+ifnull(values(%s), 0)", cexpr.colName.String())
+		}
+	}
+	return buf.ParsedQuery()
+}
+
+func (tpb *tablePlanBuilder) generateUpdateStatement() *sqlparser.ParsedQuery {
+	if tpb.onInsert == insertIgnore {
+		return tpb.generateInsertStatement()
+	}
+	bvf := &bindvarFormatter{}
+	buf := sqlparser.NewTrackedBuffer(bvf.formatter)
+	buf.Myprintf("update %v set ", tpb.name)
+	separator := ""
+	for _, cexpr := range tpb.colExprs {
+		if cexpr.isGrouped || cexpr.isPK {
+			continue
+		}
+		buf.Myprintf("%s%s=", separator, cexpr.colName.String())
+		separator = ", "
+		switch cexpr.operation {
+		case opExpr:
+			bvf.mode = bvAfter
+			buf.Myprintf("%v", cexpr.expr)
+		case opCount:
+			buf.Myprintf("%s", cexpr.colName.String())
 		case opSum:
 			buf.Myprintf("%s", cexpr.colName.String())
-			if before {
-				bvf.mode = bvBefore
-				buf.Myprintf("-ifnull(%v, 0)", cexpr.expr)
+			bvf.mode = bvBefore
```

Comment:
Comment here? On updates we want to subtract the old value and add the new value to the sum.